lib/chef/provider/package.rb

```diff
@@ -414,19 +433,19 @@ def target_version_array
             each_package do |package_name, new_version, current_version, candidate_version|
               case action
               when :upgrade
-                if version_equals?(current_version, new_version)
-                  # this is an odd use case
-                  logger.trace("#{new_resource} #{package_name} #{new_version} is already installed -- you are equality pinning with an :upgrade action, this may be deprecated in the future")
-                  target_version_array.push(nil)
-                elsif version_equals?(current_version, candidate_version)
-                  logger.trace("#{new_resource} #{package_name} #{candidate_version} is already installed")
+                if current_version.nil?
+                  logger.trace("#{new_resource} has no existing installed version. Installing install #{candidate_version}")
+                  target_version_array.push(candidate_version)
+                elsif version_equals?(current_version, new_version)
```

Comment:
Well, I downloaded this code, monkeypatched it into our code base, ran my tests, and I'm happy to report that the tests succeeded. It seems we never tried to do the aforementioned broken comparisons because `current_version` was nil (even though there was a current version of the package installed). I think you tried to explain this earlier to me when you said:

> went a completely different direction so that the subclass can direct the superclass that the installed_version ("current_version" in the superclass) is nil if it doesn't match the version constraint.  so that value can be nil if there's a package installed but it doesn't match what the resource wants.

That said, while running chef in debug mode I am still seeing a lot of `versioncompare` calls, some with version numbers that we will parse incorrectly:
```
[2020-12-10T16:46:28-08:00] DEBUG: sending '{"action":"versioncompare","versions":["0:2020.2.41-80.0.el8_2.noarch","0:2020.2.41-80.0.el8_2.noarch"]}' to python helper
[2020-12-10T16:46:34-08:00] DEBUG: sending '{"action":"versioncompare","versions":["0:60.3-2.el8_1.x86_64","0:60.3-2.el8_1.x86_64"]}' to python helper
[2020-12-10T16:46:34-08:00] DEBUG: sending '{"action":"versioncompare","versions":["0:60.3-2.el8_1.x86_64","0:60.3-2.el8_1.x86_64"]}' to python helper
[2020-12-10T16:46:34-08:00] DEBUG: sending '{"action":"versioncompare","versions":["1:0.17-73.el8_1.1.x86_64","1:0.17-73.el8_1.1.x86_64"]}' to python helper
```

Most of these seem to be occurring with multi-package resources in the :upgrade path on line 446:
```
                 elsif version_equals?(current_version, candidate_version)
```

But in all these cases we're doing an equality comparison on identical strings, so it seems like things will work even though we're not parsing EVRAs correctly. It seems like your changes essentially eliminate all version comparisons (other than for equality comparisons for equal strings). If that's true then the dnf versioncompare code is dead, so why not remove it and update the dnf 'python_helper.rb`compare_versions()' function to assert v1 == v2 and always return 0? That'd make the code less complex and the constraints more obvious. (If that assumption is not true then it seems like there should be a way to make the version comparisons fail with the current parsing code.)

Reply:
>  because current_version was nil (even though there was a current version of the package installed)

yep that's the point of `installed_version_satisfies_version_constraints` -- if it doesn't satisfy the version constraints then its nil, even though it is installed.

have to parse the rest of that tomorrow, but you're right that versioncompare is basically unimportant now.